### PR TITLE
HDDS-8681. [SNAPSHOT] Intermittent failure in TestOMDbCheckpointServlet#testWriteDbDataToStream

### PR DESCRIPTION
**Pull request by GeorgeJahad**
## What changes were proposed in this pull request?
This intermittent failure is caused by a race condition with the om double-buffer flush-thread.

the bootstrap lock here should have been preventing this problem from happening: https://github.com/apache/ozone/blob/2fbed633e87ba455e7635e75356df3ce2dd906ac/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java#L395

The reason it wasn't was because I was taking the lock after the checkpoint was created instead before, so I moved it to before and that fixed the problem.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-8681

## How was this patch tested?

Without this fix, I can make this failure happen consistently by inserting a thread sleep in the double buffer thread here:  https://github.com/GeorgeJahad/ozone/blob/15d451742a/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/snapshot/OMSnapshotCreateResponse.java#L84-L88

With the fix, it no longer fails even with the thread sleep.

### CODE REVIEW

**Comment by adoroszlai** [comment]
@hemantk-12 please review

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java [hunk 1]
```diff
@@ -885,4 +886,27 @@ public void close() {
   public long getDiffCleanupServiceInterval() {
     return diffCleanupServiceInterval;
   }
+
+  // Wait for double buffer to be flushed and confirm that the
+  // snapshot directory exists.
+  public void waitForSnapshotFlush(String volumeName, String bucketName,
```
Comment (smengcl) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java:892]: Looks like this is only used in tests. We can add an annotation for that.

```suggestion
  @VisibleForTesting
  public void waitForSnapshotFlush(String volumeName, String bucketName,
```

**Review by smengcl — APPROVED** [review]
lgtm

**Comment by hemantk-12** [comment]
Had a offline discussion with @GeorgeJahad and he is going to double if `waitForSnapshotFlush` is actually needed because tarball creation waits for double buffer to flush.

FYI: @smengcl

**Comment by GeorgeJahad** [comment]
@hemantk-12 correctly pointed out that the double buffer flush in the bootstrap lock here should have been preventing this problem from happening: https://github.com/apache/ozone/blob/2fbed633e87ba455e7635e75356df3ce2dd906ac/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OMDBCheckpointServlet.java#L395

The reason it wasn't was because I was taking the lock after the checkpoint was created instead after, so I moved it to before and that fixed the problem.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java [hunk 2]
```diff
@@ -145,28 +149,29 @@ private void generateSnapshotCheckpoint(HttpServletRequest request,
...
+    if (sstParam != null) {
+      receivedSstList.addAll(
+          Arrays.stream(sstParam)
+              .filter(s -> s.endsWith(ROCKSDB_SST_SUFFIX))
+              .distinct()
+              .collect(Collectors.toList()));
+      LOG.info("Received excluding SST {}", receivedSstList);
+    }
+
+    try  (BootstrapStateHandler.Lock lock = getBootstrapStateLock().lock()) {
```
Comment (smengcl) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java:174]:
```suggestion
    try (BootstrapStateHandler.Lock lock = getBootstrapStateLock().lock()) {
```

**Review by smengcl — COMMENTED** [review]
Thanks @GeorgeJahad . Pls take a look @hemantk-12

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java [hunk 3]
```diff
@@ -301,4 +306,27 @@ public void writeDbDataToStream(DBCheckpoint checkpoint,
     writeDBCheckpointToStream(checkpoint, destination,
         toExcludeList, excludedList);
   }
+
+  @Override
+  public BootstrapStateHandler.Lock getBootstrapStateLock() {
+    return lock;
+  }
+
+/**
```
Comment (hemantk-12) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/DBCheckpointServlet.java:315]: nit:
```suggestion
  /**
   * This lock is a no-op but can overridden by child classes.
   */
    public static class Lock extends BootstrapStateHandler.Lock {
```

**Review by hemantk-12 — APPROVED** [review]
Thanks for the patch @GeorgeJahad.

Overall looks good to me. Left a minor comment.

**Pull request merged**